Source/WebCore/platform/graphics/gstreamer/eme/CDMOpenCDM.cpp

```diff
@@ -69,46 +77,185 @@ class CDMPrivateOpenCDM : public CDMPrivate {
+    self.m_taskQueue.enqueueTask([session, method, message, messageLength]() {
+        OCDMSessionCallbacksNotifier::tryNotify(session, method, message, messageLength);
+    });
+}
+
+}  // namespace
+
+class CDMInstanceOpenCDM::Session : public ThreadSafeRefCounted<CDMInstanceOpenCDM::Session>, public OCDMSessionCallbacksNotifier::Observer {
 public:
-    static Ref<Session> create(const media::OpenCdm& source, Ref<WebCore::SharedBuffer>&& initData);
+    using DidGenerateChallengeCallback = Function<void(Session*)>;
+    using DidUpdateKeyCallback = Function<void(Session*, OCDMKeyStatus, const String&)>;
+    using DidChangeCallback = Function<void(Session*, bool success, const String&)>;
 
-    bool isValid() const { return m_isValid; }
+    static Ref<Session> create(OpenCDMAccessor&, const String&, const char*, Ref<WebCore::SharedBuffer>&&, LicenseType, Ref<WebCore::SharedBuffer>&&);
+    ~Session() override;
+
+    bool isValid() const { return m_session.get() && m_message && !m_message->isEmpty(); }
+    const String& id() const { return m_id; }
     Ref<SharedBuffer> message() const { ASSERT(m_message); return Ref<SharedBuffer>(*m_message.get()); }
     bool needsIndividualization() const { return m_needsIndividualization; }
     const Ref<WebCore::SharedBuffer>& initData() const { return m_initData; }
-    std::pair<media::OpenCdm::KeyStatus, String> update(const uint8_t* data, unsigned length);
-    std::pair<bool, String> load();
-    std::pair<bool, String> remove();
-    bool close() { return !m_session.Close(); }
-    media::OpenCdm::KeyStatus lastStatus() const { return m_lastStatus; }
+    void generateChallenge(DidGenerateChallengeCallback&&);
+    void update(const uint8_t*, unsigned, DidUpdateKeyCallback&&);
+    void load(DidChangeCallback&&);
+    void remove(DidChangeCallback&&);
+    bool close() { return !opencdm_session_close(m_session.get()); }
+    OCDMKeyStatus lastStatus() const { return m_lastStatus; }
     bool containsInitData(const InitData& initData) const {
         return m_initData->size() >= initData.sizeInBytes() && memmem(m_initData->data(), m_initData->size(), initData.characters8(), initData.sizeInBytes());
     }
 
 private:
+    using ScopedSession = std::unique_ptr<OpenCDMSession, WTF::Function<void(OpenCDMSession*)>>;
+    struct KeyUpdateData {
+        WTF_MAKE_NONCOPYABLE(KeyUpdateData);
+    public:
+        KeyUpdateData(DidUpdateKeyCallback&& callback, const uint8_t* data, const unsigned length)
+            : m_callback(WTFMove(callback))
+            , m_keySize(length)
+        {
+            m_key = std::make_unique<uint8_t[]>(m_keySize);
+            std::copy_n(data, m_keySize, m_key.get());
+        }
+        KeyUpdateData(KeyUpdateData&& other)
+        {
+            m_callback = WTFMove(other.m_callback);
+            m_key = WTFMove(other.m_key);
+        }
+        DidUpdateKeyCallback m_callback;
+        std::unique_ptr<uint8_t[]> m_key;
```

Comment:
Because you mentioned it's not a strong feeling so I took it as a suggestion and TBH I do not see any advantage of one unique ptr over the other

Reply:
`GUniquePtr` is more WebKit style and they are even the same thing with a destructor to be `g_free` instead of _delete_, but anyway, probably not worth the time.